tests/mypy_test.py

```diff
@@ -0,0 +1,89 @@
+import glob
+import os
+import re
+from subprocess import Popen, PIPE
+from typing import Text, List, Iterable, Tuple, Any
+
+import pytest
+import yaml
+
+
+DEFAULT_PATH = 'main.py'
+ENCODING = 'utf-8'
+
+
+def load_test_cases(root: Text) -> Iterable[Tuple[Text, Text, List[dict]]]:
+    paths = glob.glob(os.path.join(root, '*.yml'))
+    assert paths, 'No test data found in %s' % os.path.abspath(root)
+
+    for path in paths:
+        with open(path, 'rb') as fd:
+            raw_data = yaml.load(fd)
+        for case in raw_data:
+            _, filename = os.path.split(path)
+            name = case.get('case')
+            assert name, 'Unnamed case: %r' % case
+            files = case.get('files', [])
+            unique = {file.get('path', DEFAULT_PATH) for file in files}
+            assert len(files) == len(unique), 'Duplicate files in %s' % name
+            yield filename, name, files
+
+
+def inline_errors(errors: List[Text], files: List[dict]) -> List[dict]:
+    contents = {
+        file.get('path', DEFAULT_PATH): strip_errors(file.get('content', ''))
+        for file in files}
+
+    for error in errors:
+        index, message, path = parse_error_line(error)
+        content = contents.get(path)
+        assert content, 'Cannot find contents of %s' % path
+        lines = content.splitlines(keepends=True)
+        assert 0 <= index < len(lines)
+        line = lines[index]
+        updated = '%s# E: %s\n' % (line.rstrip('\n'), message)
+        lines[index] = updated
+        contents[path] = ''.join(lines)
+
+    results = [file.copy() for file in files]
+    for result in results:
+        result['content'] = contents.get(result.get('path', DEFAULT_PATH), '')
+    return results
+
+
+def strip_errors(s: Text) -> Text:
+    return re.sub(r'#\s*E:.*', r'', s)
+
+
+def parse_error_line(line: Text) -> Tuple[int, Text, Text]:
+    m = re.match(r'(.+):([0-9]+): (error): (.+)', line)
+    assert m, 'Cannot parse output line: %s' % line
+    path = m.group(1)
+    index = int(m.group(2)) - 1
+    message = m.group(4)
+    return index, message, path
+
+
+@pytest.mark.parametrize('filename,case,files', load_test_cases('test-data'))
+def test_yaml_case(filename: Text, case: Text, files: List[dict],
```

Comment:
Pytest allows easily writing a plugin to directly parse non-Python files as pytest-native test files containing individual tests, rather than using test parametrization like this. I've done this before, and it works very well and has a number of nice properties, e.g.:

- output when running tests has a line of dots per actual test file, not just a single line of dots referring uselessly to this file
- test names in failure output can be exactly the name listed in the yaml file, with no repeated `test_yaml_case` cruft
- you can use the obvious `pytest path/to/test-basic.yaml` natively, to run the tests in `test-basic.yaml` only, just like you would with a file containing Python tests

Sample code for this approach can be found at https://docs.pytest.org/en/latest/example/nonpython.html